game-engine/core/src/main/java/za/co/entelect/challenge/core/engine/TowerDefenseRoundProcessor.java

```diff
@@ -103,24 +103,26 @@ private static boolean positionMatch(Cell a, Cell b) {
     }
 
     private void calculateMissileMovement() {
-
         towerDefenseGameMap.getMissiles()
-                .forEach(missile -> IntStream.rangeClosed(1, missile.getSpeed()) // higher speed bullets
-                        .forEach(i -> {
-                            try {
-                                towerDefenseGameMap.moveMissileSingleSpace(missile);
-                                towerDefenseGameMap.getBuildings().stream()
-                                        .filter(b -> b.isConstructed() && positionMatch(missile, b))
-                                        .findAny()
-                                        .ifPresent(b -> {
-                                            b.damageSelf(missile);
-                                            towerDefenseGameMap.getPlayerByStream(missile.getPlayerType())
-                                                    .forEach(player -> player.addScore(b.getDestroyMultiplier() * missile.getDamage()));
-                                        });
-                            } catch (Exception e) {
-                                e.printStackTrace();
-                            }
-                        })
+                .forEach(missile ->
+                        IntStream.rangeClosed(1, missile.getSpeed()) // higher speed bullets
+                                .forEach(i -> {
+                                    towerDefenseGameMap.moveMissileSingleSpace(missile);
+                                    towerDefenseGameMap.getBuildings().stream()
+                                            .filter(b -> b.isConstructed()
+                                                    && positionMatch(missile, b)
+                                                    && !b.isPlayers(missile.getPlayerType())
+                                                    && b.getHealth() > 0)
+                                            .forEach(b -> {
+                                                TowerDefensePlayer missileOwner = null;
+                                                try {
+                                                    missileOwner = towerDefenseGameMap.getPlayer(missile.getPlayerType());
+                                                } catch (Exception e) {
+                                                    e.printStackTrace();
+                                                }
+                                                b.damageSelf(missile, missileOwner);
```

Comment:
&& _!b.isPlayers(missile.getPlayerType())_
building is not owned by the missile owner

_&& b.getHealth() > 0_
only check alive buildings